### PR TITLE
Fix docstring bug

### LINKED ISSUE #11194: Natspec does not allow tags on non-public state variables
## Description

[Release notes of Solidity 0.8.2](https://github.com/ethereum/solidity/releases/tag/v0.8.2) announces following:

> Support field `documentation` to hold NatSpec comments above each statement.

While this is working for most of the statements, the following sample is not able to be compiled:

```solidity
/// SPDX-License-Identifier: UNLICENSED

contract Test {
    function some() public {
        /// VariableDeclarationStatement documentation
        uint x = 1;
    }
}
```

The related error:
```json
{
    "errors": [
        {
            "component": "general",
            "errorCode": "2837",
            "formattedMessage": "ParserError: Only state variables or file-level variables can have a docstring.\n --> sample.sol:6:14:\n  |\n6 |         uint x = 1;\n  |              ^\n\n",
            "message": "Only state variables or file-level variables can have a docstring.",
            "severity": "error",
            "sourceLocation": {
                "end": 155,
                "file": "sample.sol",
                "start": 154
            },
            "type": "ParserError"
        }
    ],
    "sources": {}
}

```

However there is a workaround to use single-tuple syntax, that is producing valid output:

```solidity
/// SPDX-License-Identifier: UNLICENSED

contract Test {
    function some() public {
        /// VariableDeclarationStatement documentation
        (uint x) = (1);
    }
}
```

## Environment

- [SolcJS package v0.8.3 at NPM](https://www.npmjs.com/package/solc/v/0.8.3) (current latest)

```bash
$ node --version
v12.18.3

$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.1 LTS
Release:	20.04
Codename:	focal
```

## Steps to Reproduce

**package.json**
```json
{
  "name": "solc-js-083-test",
  "version": "0.0.0",
  "main": "index.js",
  "dependencies": {
    "solc-0.8.3": "npm:solc@0.8.3"
  }
}
```

**sample.sol**
```solidity
/// SPDX-License-Identifier: UNLICENSED

contract Test {
    function some() public {
        /// VariableDeclarationStatement documentation
        uint x = 1;
    }
}
```

**index.js**
```javascript
const fs = require("fs");
const compiler = require("solc-0.8.3");

const fileName = "sample.sol";
const content = fs.readFileSync(fileName, { encoding: "utf-8" });

const input = {
    language: "Solidity",
    sources: {
        [fileName]: {
            content
        }
    },
    settings: {
        outputSelection: {
            "*": {
                "*": ["*"],
                "": ["*"]
            }
        }
    }
};

const output = compiler.compile(JSON.stringify(input), {});
const parsed = JSON.parse(output);

console.log(JSON.stringify(parsed, undefined, 4));
```

**Run command**
```bash
npm install # run once prior to execution to setup dependencies
node index.js
```

Regards.

### PR DESCRIPTION
**Pull request by hrkrshnn**
Closes https://github.com/ethereum/solidity/issues/11194

### CODE REVIEW

File: test/libsolidity/ASTJSON/documentation_local_variable.sol [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	}
+	function g(
+		/// documentation for param1
+		uint param1,
+		/// documentation for param2
+		uint param2,
+		/// documentation for param3
+		uint param3
+	) public {}
+}
```
Comment (hrkrshnn) [inline test/libsolidity/ASTJSON/documentation_local_variable.sol:23]: In this test, only the documentation for `uint x = 1` and `uint j = 0` shows up in the json file.

File: libsolidity/analysis/DocStringTagParser.cpp [hunk 2]
```diff
@@ -71,7 +71,7 @@ bool DocStringTagParser::visit(VariableDeclaration const& _variable)
 		if (_variable.isPublic())
 			parseDocStrings(_variable, _variable.annotation(), {"dev", "notice", "return", "inheritdoc"}, "public state variables");
 		else
-			parseDocStrings(_variable, _variable.annotation(), {"dev", "inheritdoc"}, "non-public state variables");
+			parseDocStrings(_variable, _variable.annotation(), {"dev", "notice", "inheritdoc"}, "non-public state variables");
```
Comment (chriseth) [inline libsolidity/analysis/DocStringTagParser.cpp:74]: I don't think notice is useful for non-public state variables

File: test/libsolidity/ASTJSON/documentation_local_variable.sol (hunk 1 above)
Comment (chriseth) [reply test/libsolidity/ASTJSON/documentation_local_variable.sol:23]: Can you check why?

Comment (hrkrshnn) [reply test/libsolidity/ASTJSON/documentation_local_variable.sol:23]: We need to export `StructuredDocumentation` for this to show up. See https://gitter.im/ethereum/solidity-dev?at=6065e3aad765936399ca3939.

File: test/libsolidity/ASTJSON/documentation_local_variable.sol [hunk 3]
```diff
@@ -0,0 +1,25 @@
+contract C {
+    function f() public pure {
+        /// Documentation for x; will appear in ast json
+        uint x = 1;
```
Comment (hrkrshnn) [inline test/libsolidity/ASTJSON/documentation_local_variable.sol:4]: The parse only JSON would already contain "documentation for x, ..." even without this fix.

File: libsolidity/analysis/DocStringTagParser.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/analysis/DocStringTagParser.cpp:74]: Should it be removed?

**Comment by chriseth** [comment]
@hrkrshnn we discussed this last week, I don't remember - what was the conclusion?

**Comment by hrkrshnn** [comment]
@chriseth I'm not sure if we had a conclusion.

I would say the following should be how we resolve this:

- Allow non-public variable declarations to have NatSpec, and therefore `notice` since it's the default. (So local variables and non-public state variables) Already implemented here.
- Export `StructuredDocumentation` instead of `Documented` for statements (https://github.com/ethereum/solidity/pull/11205#discussion_r605759251) should be done separately. Maybe we can do this as a bugfix, since the changelog already mentions this https://github.com/ethereum/solidity/blob/develop/Changelog.md#082-2021-03-02.

File: libsolidity/analysis/DocStringTagParser.cpp (hunk 2 above)
Comment (hrkrshnn) [reply libsolidity/analysis/DocStringTagParser.cpp:74]: @leonardoalt this was mentioned in the chat on thursday. The problem with not allowing `notice` here is:

```solidity
contract Test {
	/// this is an error
	uint y = 1;
	function some() public {
		/// this is not an error
		uint x = 1;
	}
}
```
I find this behaviour weird and would prefer allowing notice everywhere.

**Comment by chriseth** [comment]
And what is the problem that prevents us from making dev the default for non-public variables?

**Comment by hrkrshnn** [comment]
@chriseth Making "notice" as default for non-public variables is that currently it's done in DocStringParser.

https://github.com/ethereum/solidity/blob/develop/libsolidity/parsing/DocStringParser.cpp#L112

While it's probably possible to get the information that the node is a public state variable, I'm not sure if it will run into issues. Let me implement it and see how it goes.

File: test/libsolidity/syntaxTests/constants/constant_natspec_user.sol [hunk 4]
```diff
@@ -1,4 +1,3 @@
 /// Documentation
 uint constant x = 8;
 // ----
-// DocstringParsingError 6546: (0-18): Documentation tag @notice not valid for file-level variables.
```
Comment (hrkrshnn) [inline test/libsolidity/syntaxTests/constants/constant_natspec_user.sol]: Because this is now `dev`, the error disappears.

**Comment by chriseth** [comment]
I'm sorry, it seems I misunderstood: Do we currently default to `@notice` for everything, even for internal functions, contracts and statements inside functions?

**Comment by hrkrshnn** [comment]
@chriseth I reverted the change to `dev`. Now, it allows `notice` on non-public state variables. And removes a parser check about `public` / `non-public` for NatSpec.

**Review by chriseth — APPROVED** [review]

**Pull request merged**